aws/resource_aws_kinesis_analytics_application.go

```diff
@@ -0,0 +1,1521 @@
+				addOpts := &kinesisanalytics.AddApplicationOutputInput{
+					ApplicationName:             aws.String(name),
+					CurrentApplicationVersionId: aws.Int64(int64(version)),
+					Output: output,
+				}
+				err := resource.Retry(1*time.Minute, func() *resource.RetryError {
+					_, err := conn.AddApplicationOutput(addOpts)
+					if err != nil {
+						if isAWSErr(err, kinesisanalytics.ErrCodeInvalidArgumentException, "Kinesis Analytics service doesn't have sufficient privileges") {
+							return resource.RetryableError(err)
+						}
+						return resource.NonRetryableError(err)
+					}
+					return nil
+				})
+				if err != nil {
+					return fmt.Errorf("Unable to add application outputs: %s", err)
+				}
+				version = version + 1
+			}
+		}
+	}
+
+	oldReferenceData, newReferenceData := d.GetChange("reference_data_sources")
+	if len(oldReferenceData.([]interface{})) == 0 && len(newReferenceData.([]interface{})) > 0 {
+		if v := d.Get("reference_data_sources").([]interface{}); len(v) > 0 {
+			for _, r := range v {
+				rd := r.(map[string]interface{})
+				referenceData := createReferenceData(rd)
+				addOpts := &kinesisanalytics.AddApplicationReferenceDataSourceInput{
+					ApplicationName:             aws.String(name),
+					CurrentApplicationVersionId: aws.Int64(int64(version)),
+					ReferenceDataSource:         referenceData,
+				}
+				err := resource.Retry(1*time.Minute, func() *resource.RetryError {
+					_, err := conn.AddApplicationReferenceDataSource(addOpts)
+					if err != nil {
+						if isAWSErr(err, kinesisanalytics.ErrCodeInvalidArgumentException, "Kinesis Analytics service doesn't have sufficient privileges") {
+							return resource.RetryableError(err)
+						}
+						return resource.NonRetryableError(err)
+					}
+					return nil
+				})
+				if err != nil {
+					return fmt.Errorf("Unable to add application reference data source: %s", err)
+				}
+				version = version + 1
+			}
+		}
+	}
+
+	return resourceAwsKinesisAnalyticsApplicationRead(d, meta)
+}
+
+func resourceAwsKinesisAnalyticsApplicationDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).kinesisanalyticsconn
+	name := d.Get("name").(string)
+	createTimestamp, parseErr := time.Parse(time.RFC3339, d.Get("create_timestamp").(string))
+	if parseErr != nil {
+		return parseErr
+	}
+
+	log.Printf("[DEBUG] Kinesis Analytics Application destroy: %v", d.Id())
+	deleteOpts := &kinesisanalytics.DeleteApplicationInput{
+		ApplicationName: aws.String(name),
+		CreateTimestamp: aws.Time(createTimestamp),
+	}
+	_, deleteErr := conn.DeleteApplication(deleteOpts)
+	if deleteErr != nil {
+		return deleteErr
+	}
+
+	log.Printf("[DEBUG] Kinesis Analytics Application deleted: %v", d.Id())
+	return nil
+}
+
+func createCloudwatchLoggingOption(clo map[string]interface{}) *kinesisanalytics.CloudWatchLoggingOption {
+	cloudwatchLoggingOption := &kinesisanalytics.CloudWatchLoggingOption{
+		LogStreamARN: aws.String(clo["log_stream"].(string)),
```

Comment:
We should prefer to stick to the API (and more clear) attribute naming for any of these `ARN` parameters that accept an ARN, e.g. `log_stream_arn`

This applies to a large amount of the attributes below as well so I'll withhold from commenting on each of them.